File: NET/JNetHotel.Sharp/Natives.cs
using System;
using System.Runtime.InteropServices;
using JNetHotel.Sharp.API;
using JNetHotel.Sharp.BSD;
using JNetHotel.Sharp.Linux;
using JNetHotel.Sharp.Mac;
using JNetHotel.Sharp.Windows;

namespace JNetHotel.Sharp
{
    public static class Natives
    {
        public static IVmRef GetVmRef()
        {
            if (IsLinux)
                return new LinuxVmRef();
            if (IsWindows)
                return new WinVmRef();
            if (IsMac)
                return new MacVmRef();
            if (IsBSD)
                return new BsdVmRef();

            throw new InvalidOperationException(RuntimeInformation.OSDescription);
        }

        private static bool IsWindows => RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
        private static bool IsLinux => RuntimeInformation.IsOSPlatform(OSPlatform.Linux);
        private static bool IsMac => RuntimeInformation.IsOSPlatform(OSPlatform.OSX);
        private static bool IsBSD => RuntimeInformation.IsOSPlatform(OSPlatform.FreeBSD);
    }
}

File: JVM/jnethotel-java/src/main/java/jnethotel/java/Natives.java
package jnethotel.java;

import com.sun.jna.Platform;
import jnethotel.java.api.IVmRef;
import jnethotel.java.linux.LinuxVmRef;
import jnethotel.java.mac.MacVmRef;
import jnethotel.java.windows.WinVmRef;

public final class Natives {

    public static IVmRef getVmRef() {
        if (isLinux())
            return new LinuxVmRef();
        if (isWindows())
            return new WinVmRef();
        if (isMac())
            return new MacVmRef();

        var desc = System.getProperty("os.name") + " " + System.getProperty("os.version");
        throw new UnsupportedOperationException(desc.trim());
    }

    private static boolean isWindows() {
        return Platform.isWindows();
    }

    private static boolean isLinux() {
        return Platform.isLinux();
    }

    private static boolean isMac() { return Platform.isMac(); }
}
